src/node_contextify.cc

```diff
@@ -1637,22 +1455,43 @@ static void CompileFunctionForCJSLoader(
 
   Local<Function> fn;
   if (!maybe_fn.ToLocal(&fn)) {
-    if (try_catch.HasCaught() && !try_catch.HasTerminated()) {
-      errors::DecorateErrorStack(env, try_catch);
-      if (!try_catch.HasTerminated()) {
-        try_catch.ReThrow();
-      }
-      return;
-    }
+    return scope.EscapeMaybe(MaybeLocal<Function>());
   }
 
-  bool cache_rejected = false;
 #ifndef DISABLE_SINGLE_EXECUTABLE_APPLICATION
   if (used_cache_from_sea) {
-    cache_rejected = source.GetCachedData()->rejected;
+    *cache_rejected = source.GetCachedData()->rejected;
   }
 #endif
 
+  return scope.Escape(fn);
+}
+
+static void CompileFunctionForCJSLoader(
+    const FunctionCallbackInfo<Value>& args) {
+  CHECK(args[0]->IsString());
+  CHECK(args[1]->IsString());
+  Local<String> code = args[0].As<String>();
+  Local<String> filename = args[1].As<String>();
+  Isolate* isolate = args.GetIsolate();
+  Local<Context> context = isolate->GetCurrentContext();
+  Environment* env = Environment::GetCurrent(context);
+
+  bool cache_rejected = false;
+  Local<Function> fn;
+  {
+    TryCatchScope try_catch(env);
+    if (!CompileFunctionForCJSLoader(
+             env, context, code, filename, &cache_rejected)
+             .ToLocal(&fn)) {
+      CHECK(try_catch.HasCaught());
+      CHECK(!try_catch.HasTerminated());
+      errors::DecorateErrorStack(env, try_catch);
+      try_catch.ReThrow();
+      return;
+    }
+  }
```

Comment:
> if it’s entry point, and the detection flag is enabled, and it looks like ESM, use the path that allow TLA, and ignore the exports because uh no one needs the `module.exports` of the entry point

What if the entry point gets required or imported a second time? Like `node entry.js` imports `foo.js` which imports `entry.js`? The second import would need the exports from `entry.js`.

Can we land your follow-up and unflag `detect-module` while `require(esm)` is still flagged? Or is `require(esm)` possibly close enough to being ready to unflag that we could just do them together?

Reply:
> The second import would need the exports from entry.js.

If entry.js is ESM, then that's the namespace, not `module.exports`, and we'll return the namespace of the cached module in the ESM loader, because import is not handled by CJS loader.

Reply:
> Can we land your follow-up and unflag detect-module while require(esm) is still flagged?

I don't know for sure, locally my followup does pass `test/parallel/test-debugger-exceptions.js` even when detect-module is unflagged, though, if you think that's the last blocker. `require(esm)` will need to be flagged for at least 22, blockers are `__esModule` interop and customization hooks support, and maybe there are more as we get more feedback from 22.

Reply:
> In general I'm not too comfortable with just unflagging an experimental feature that really needs some battle testing from users at least.

Especially for a feature like this one. As soon as it can be used without a flag, people will start to publish packages that depend on it. Then we will be blamed if we break them.